Trigger custom DOM event when layout changes (#5555)
hypothesis/client: docs/publishers/events.rst

Author: I considered it, but I thought the events section could grow enough to be a page of its own.

Events
======

> Hypothesis client emits a few custom DOM events at ``document.body``, to which your app can react.

Reviewer: ```suggestion
The Hypothesis client emits custom DOM events at ``document.body``, to which your app can react.
```

.. option:: hypothesis:layoutchange

   This event is emitted when the sidebar layout changes. eg. when it is opened or closed.

>    It extends `CustomEvent <https://developer.mozilla.org/docs/Web/API/CustomEvent>`_, so its payload can be found inside the ``detail`` property.

Reviewer: This applies to all events, not just this event, so it should either be part of the introduction or a separate paragraph after the list of events.

Properties
----------

.. option:: sidebarLayout

  .. option:: expanded

     ``Boolean`` True if the sidebar is open

  .. option:: width

     ``Number`` The width of the sidebar in pixels

  .. option:: height

     ``Number`` The height of the sidebar in pixels

.. option:: sideBySideActive

  ``Boolean`` Indicates whether side-by-side mode is active

.. code-block:: javascript

    document.body.addEventListener('hypothesis:layoutchange', event => {
      console.log(event.detail.sidebarLayout.expanded);
      console.log(event.detail.sidebarLayout.width);
      console.log(event.detail.sidebarLayout.height);

>       console.log(event.detail.sidebarLayout.toolbarWidth);

Reviewer: This line is obsolete

Custom DOM events
=================

> Hypothesis client emits a few custom DOM events to which your app can react.

Reviewer: Related to this, there is an existing issue about documenting the `scrolltorange` event. Before we do that though, we should change it to have a `hypothesis:` prefix. We may also need to keep the old event for backwards compatibility for a period of time.

.. option:: hypothesis:layoutchange

>    This event is emitted when the sidebar layout changes. Changes to the layout occur on load, when the sidebar is

Reviewer: ```suggestion
   This event is emitted when the sidebar layout changes. eg. when it is opened or closed.
```

I don't think we should list all the events that can trigger this, since they may change in future.

   toggled to show and hide, and when the user adjusts the sidebar manually.

>    The event will wrap some properties that can be used to tell the sidebar status and size.

Reviewer: This sentence seems unnecessary to me. I think the documentation that follows is self-explanatory enough. A heading like "Properties" would make more sense, plus a note that events extend `CustomEvent` and properties are available on `detail`.

   .. option:: sidebarLayout

      .. option:: expanded

>          ``Boolean`` If the sidebar is open, this value will be true.

Reviewer: ```suggestion
         ``Boolean`` True if the sidebar is open
```

      .. option:: width

>          ``Number`` The current visible height of the sidebar.

Reviewer: ```suggestion
         ``Number`` The width of the sidebar
```

Reviewer: Width/height are mixed up here.

Reviewer 2: Do we need to clarify that these are pixel units?

Author: OMG, what did I do here 😅

      .. option:: height

>          ``Number`` The current visible width of the sidebar.

Reviewer: ```suggestion
         ``Number`` The height of the sidebar
```

      .. option:: toolbarWidth

>          ``Number`` The current visible width of the toolbar.

Reviewer: ```suggestion
         ``Number`` The width of the toolbar along the edge of the sidebar
```

   .. option:: isSideBySideActive

>       ``Boolean`` Tells if the side-by-side mode is active.

Reviewer: See my comments elsewhere about renaming this to `sideBySideActive`.

Reviewer 2: I also agree with `sideBySideActive` vs. `isSideBySideActive`

Custom DOM events
=================

> Hypothesis client emits a few custom DOM events to which your app can react.

Reviewer: The documentation doesn't specify which target DOM events are dispatched at, although the code example below indicates the document body. I suggest we just specify the document body, although internally this is abstracted from the `Guest` code via the `Guest.element` property.
